plugins/experimental/traffic_dump/traffic_dump.cc

```diff
@@ -315,12 +348,160 @@ session_aio_handler(TSCont contp, TSEvent event, void *edata)
     return TS_SUCCESS;
   }
   default:
-    TSDebug(PLUGIN_NAME, "session_aio_handler(): unhandled events %d", event);
+    TSDebug(PLUGIN_NAME, "%s: unhandled events %d", __func__, event);
     return TS_ERROR;
   }
   return TS_SUCCESS;
 }
 
+std::string
+request_body_get(TSHttpTxn txnp)
+{
+  TSDebug(PLUGIN_NAME, "getting the body");
+  TSIOBufferReader post_buffer_reader = TSHttpTxnPostBufferReaderGet(txnp);
+  int64_t read_avail                  = TSIOBufferReaderAvail(post_buffer_reader);
+  if (read_avail == 0) {
+    TSIOBufferReaderFree(post_buffer_reader);
+    return {};
+  }
+
+  // std::string has no reserving allocator. So create an explicitly empty
+  // string then reserve the bytes for later population.
+  std::string body_string;
+  body_string.resize(read_avail);
+  TSIOBufferReaderCopy(post_buffer_reader, body_string.data(), read_avail);
+  TSIOBufferReaderFree(post_buffer_reader);
+
+  TSDebug(PLUGIN_NAME, "returning %ld body bytes", body_string.size());
+  return body_string;
+}
+
+/** Print the json data for a transaction
+ *
+ * @param[in] txnp The reference to the transaction.
+ * @param[in] ssnData The plugin's data for the session.
+ * @param[in] txnData The plugin's data for the transaction.
+ */
+void
+print_transaction_content(TSHttpTxn txnp, SsnData *ssnData, TxnData *txnData)
+{
+  // Get UUID
+  char uuid[TS_CRUUID_STRING_LEN + 1];
+  TSAssert(TS_SUCCESS == TSClientRequestUuidGet(txnp, uuid));
+
+  // Generate per transaction json records
+  std::string txn_info;
+  // Cut down on allocations as the string is constructed piecemeal over the
+  // life of this function.
+  txn_info.reserve(4 * 1024);
+  if (!ssnData->first) {
+    txn_info += ",";
+  }
+  ssnData->first = false;
+
+  // "uuid":(string)
+  txn_info += "{" + json_entry("uuid", uuid, strlen(uuid));
+
+  // "connect-time":(number)
+  TSHRTime start_time;
+  TSHttpTxnMilestoneGet(txnp, TS_MILESTONE_UA_BEGIN, &start_time);
+  txn_info += ",\"start-time\":" + std::to_string(start_time);
+
+  // client/proxy-request/response headers
+  TSMBuffer buffer;
+  TSMLoc hdr_loc;
+  if (TS_SUCCESS == TSHttpTxnClientReqGet(txnp, &buffer, &hdr_loc)) {
+    TSDebug(PLUGIN_NAME, "Found client request");
+    if (txnData) {
+      txn_info += R"(,"client-request":)" + collect_headers(buffer, hdr_loc, txnData->request_body);
+    } else {
+      txn_info += R"(,"client-request":)" + collect_headers(buffer, hdr_loc, TSHttpTxnClientReqBodyBytesGet(txnp));
+    }
+    TSHandleMLocRelease(buffer, TS_NULL_MLOC, hdr_loc);
+    buffer = nullptr;
```

Comment:
Why clear this?

Reply:
It was there before, I just copied it over. It shouldn't be strictly necessary. I'll remove it.